### PR TITLE
[metadata] Cache GenericParamInfo for anonymous gparams in MonoImage

### PR DESCRIPTION
**Pull request by lambdageek**
### Fix Leaks of `MonoGenericInst` and `MonoGenericParam`

Fix a couple of memory leaks that show up in pathological examples like https://bugzilla.xamarin.com/show_bug.cgi?id=48707 where we repeatedly allocate provisional metadata that cause `MonoImage` mempools to grow forever.

1. The argument to `mono_metadata_get_generic_inst` should be malloc'd instead of being mempool-alloc'd since `mono_metadata_get_generic_inst` will create a copy when it makes the canonical `MonoGenericInst`.
2. When parsing an anonymous generic param, mempool-alloc at most one copy of an anonymous `MonoGenericParam` per image instead of allocating a fresh one every time.  (This matters if we repeatedly call `method_from_methodspec` when JITing one generic method that calls another.)

### Simplify `MonoGenericParam` representations and `MonoClass` caching

Previously we had `MonoGenericParam` and `MonoGenericParamFull`.  The "full" param included an extra `MonoGenericParamInfo` field.  Params without the info were "small".   Anonymous gparams were previously small.  Because small params don't have a `MonoGenericParamInfo:pklass` field, we couldn't cache the `MonoClass` created for an anonymous generic param inside the param itself.  It's not clear if this brought any space savings.

After this PR, we have `typedef MonoGenericParam MonoGenericParamFull` - all gparams are "full".
This means we can get rid of the 6 `MonoClass` caches in `MonoImage`.

### CODE REVIEW

File: mono/metadata/metadata.c [hunk 1]
```diff
@@ -3471,6 +3471,69 @@ get_anonymous_container_for_image (MonoImage *image, gboolean is_mvar)
...
+		return gparam;
+	// Create a candidate generic param and try to insert it in the cache.
+        // If multiple threads both try to publish the same param, all but one
+        // will leak, but that's okay.
+	gparam = mono_image_alloc0 (image, sizeof (MonoGenericParam));
+	gparam->owner = container;
+	gparam->num = param_num;
+
+	gparam = publish_anon_gparam (image, gparam);
+	return gparam;
```
Comment (vargaz) [inline mono/metadata/metadata.c]: It could be just return publish_...

**Comment by lambdageek** [comment]
@monojenkins build

**Comment by lambdageek** [comment]
@vargaz @kumpera Can I get a re-review - in particular the latest 3 commits change how we cache a `MonoClass` for each `MonoGenericParam` and also make all gparams into `MonoGenericParamFull`.

File: mono/metadata/metadata.c [hunk 2]
```diff
@@ -3466,6 +3474,100 @@ get_anonymous_container_for_image (MonoImage *image, gboolean is_mvar)
...
+ * LOCKING: takes the image lock.
+ */
+static MonoGenericParam*
+publish_anon_gparam_slow (MonoImage *image, MonoGenericParam *gparam)
+{
+	MonoConcurrentHashTable **cache = gparam->owner->is_method ? &image->mvar_gparam_cache : &image->var_gparam_cache;
+	if (!*cache) {
+		mono_image_lock (image);
+		if (!*cache) {
+			*cache = mono_conc_hashtable_new ((GHashFunc)mono_metadata_generic_param_hash,
```
Comment (kumpera) [inline mono/metadata/metadata.c]: This is wrong, this must use a release store.

File: mono/metadata/class-internals.h [hunk 3]
```diff
@@ -1069,7 +1052,8 @@ mono_metadata_has_generic_params (MonoImage *image, guint32 token);
 
 MONO_API MonoGenericContainer *
 mono_metadata_load_generic_params (MonoImage *image, guint32 token,
```
Comment (kumpera) [inline mono/metadata/class-internals.h]: This function is MONO_API. Do you know the reason? Did you check whether the SDKs use it?

Comment (lambdageek) [reply mono/metadata/class-internals.h]: git grep in XA, XM/XI, VSfM didn't find anything.

**Pull request merged**